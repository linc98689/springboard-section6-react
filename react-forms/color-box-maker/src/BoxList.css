.BoxList-list{
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    align-items: center;
}
.BoxList-subtitle{
    text-align: left;
    margin-left:1em;
}
.BoxList-form{
    display:flex;
    flex-direction: column;
    align-items:flex-start;
}
.BoxList-form-element{
    margin:0.5em 1em;
    display:block;
}
.BoxList-form-label{
    padding: 0.5em 0.8em;
    font-weight: bold;
}
.BoxList-form-button{
    margin: 0.5em 2em;
    background-color:rgb(212, 206, 206);
    padding: 0.5em 0.8em;
    border: 1px solid rgb(65, 64, 64);
    border-radius: 5px;
    font-weight: bold;
}