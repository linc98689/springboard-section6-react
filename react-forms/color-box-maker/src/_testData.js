const testBoxes = [
    {color:"c1", width: 1, height:1},
    {color:"c2", width: 2, height:2},
    {color:"c3", width: 3, height:3},
    {color:"c4", width: 4, height:5}
];
const tbox = {id:0, color:"c", width:1, height:1 ,
    deleteBox(id){
        let btn = document.getElementById(id);
        btn.remove();
    }};

export {testBoxes, tbox};