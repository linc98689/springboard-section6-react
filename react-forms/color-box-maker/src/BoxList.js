import React from 'react';
import {useState} from "react";
import {v4 as uuid} from "uuid";
import "./BoxList.css";
import Box from "./Box";

const INIT_BOXES = [
    {color: "black", width: "100", height: "100"},
    {color: "blue", width: "150", height: "100"},
    {color: "red", width: "200", height: "120"},
];
const COLORS = ["aqua", "black", "blue", "fuchsia", "gray", "green", "lime", "maroon", 
"navy", "olive", "purple", "red", "silver", "teal", "white", "yellow"];

const BoxList = ()=>{
    //states
    const [boxes, setBoxes] = useState(INIT_BOXES);
    const [formData, setFormData] = useState({color:"", width:"", height:""});

    // useEffect(()=>{
    //     console.log(formData);
    // }, [formData]);

    const handleChange = (evt)=>{
        const {id, value} = evt.target;
        setFormData(data =>{
            return {...data, [id]: value};
        })
    };

    const handleSubmit = (evt)=>{
        evt.preventDefault();
        setBoxes(data => [...data, formData]);
        setFormData(data =>{
            return {color:"", width:"", height:""}});
    }

    const deleteBox = (idx)=>{
        setBoxes(data => data.toSpliced(idx, 1));
    };

    // return
    return (
        <div className="BoxList">
            <h1 className="BoxList-title">Color Box Maker</h1>
            <h2 className="BoxList-subtitle">Make A New Box</h2>
            <form className="BoxList-form">
                <div className="BoxList-form-element">
                    <label htmlFor='width' className="BoxList-form-label" >Width in PX:</label>
                    <input id="width" type="text" className="BoxList-input-text" value ={formData.width} onChange={handleChange}/>
                </div>
                <div className="BoxList-form-element" >
                    <label htmlFor='height' className="BoxList-form-label" >Height in PX:</label>
                    <input id="height" type="text" className="BoxList-input-text"  value={formData.height} onChange={handleChange}/>
                </div>
                <div className="BoxList-form-element">
                    <label htmlFor="color" className="BoxList-form-label" >COLOR: </label>
                    <select id="color" onChange={handleChange}>
                        <option value="">--- Please select a color---</option>
                        {COLORS.map(c => <option value={c} key={uuid()}>{c}</option>)}
                    </select>
                </div>
                <button type="submit" onClick={handleSubmit} className="BoxList-form-element BoxList-form-button">Add</button>
            </form>
            
            <h2 className="BoxList-subtitle">Boxes You Have Made</h2>
            <div className="BoxList-list">
                {boxes.map( ({color, width, height}, idx) => <Box color={color} width={width+"px"} height={height+"px"}
                 key={uuid()} deleteBox={deleteBox} id={idx}/> )}
            </div>
        </div>
    );
}

export default BoxList;