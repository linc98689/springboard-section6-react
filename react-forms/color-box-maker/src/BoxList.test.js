import {render,  fireEvent, screen} from "@testing-library/react";
import '@testing-library/jest-dom';
import BoxList from "./BoxList";
import {testBoxes} from "./_testData";

// smoke test
test("should render", ()=>{
    render(<BoxList />);
})

// snap-shot test
test("should not change", ()=>{
    const {asFragment} = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
})

// test form submission

