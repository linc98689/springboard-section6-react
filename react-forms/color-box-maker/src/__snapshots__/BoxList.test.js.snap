// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not change 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <h1
      class="BoxList-title"
    >
      Color Box Maker
    </h1>
    <h2
      class="BoxList-subtitle"
    >
      Make A New Box
    </h2>
    <form
      class="BoxList-form"
    >
      <div
        class="BoxList-form-element"
      >
        <label
          class="BoxList-form-label"
          for="width"
        >
          Width in PX:
        </label>
        <input
          class="BoxList-input-text"
          id="width"
          type="text"
          value=""
        />
      </div>
      <div
        class="BoxList-form-element"
      >
        <label
          class="BoxList-form-label"
          for="height"
        >
          Height in PX:
        </label>
        <input
          class="BoxList-input-text"
          id="height"
          type="text"
          value=""
        />
      </div>
      <div
        class="BoxList-form-element"
      >
        <label
          class="BoxList-form-label"
          for="color"
        >
          COLOR: 
        </label>
        <select
          id="color"
        >
          <option
            value=""
          >
            --- Please select a color---
          </option>
          <option
            value="aqua"
          >
            aqua
          </option>
          <option
            value="black"
          >
            black
          </option>
          <option
            value="blue"
          >
            blue
          </option>
          <option
            value="fuchsia"
          >
            fuchsia
          </option>
          <option
            value="gray"
          >
            gray
          </option>
          <option
            value="green"
          >
            green
          </option>
          <option
            value="lime"
          >
            lime
          </option>
          <option
            value="maroon"
          >
            maroon
          </option>
          <option
            value="navy"
          >
            navy
          </option>
          <option
            value="olive"
          >
            olive
          </option>
          <option
            value="purple"
          >
            purple
          </option>
          <option
            value="red"
          >
            red
          </option>
          <option
            value="silver"
          >
            silver
          </option>
          <option
            value="teal"
          >
            teal
          </option>
          <option
            value="white"
          >
            white
          </option>
          <option
            value="yellow"
          >
            yellow
          </option>
        </select>
      </div>
      <button
        class="BoxList-form-element BoxList-form-button"
        type="submit"
      >
        Add
      </button>
    </form>
    <h2
      class="BoxList-subtitle"
    >
      Boxes You Have Made
    </h2>
    <div
      class="BoxList-list"
    >
      <div
        class="Box"
      >
        <div
          class="Box-box"
          style="background-color: black; width: 100px; height: 100px;"
        />
        <button
          class="Box-delete"
          id="0"
        >
          X
        </button>
      </div>
      <div
        class="Box"
      >
        <div
          class="Box-box"
          style="background-color: blue; width: 150px; height: 100px;"
        />
        <button
          class="Box-delete"
          id="1"
        >
          X
        </button>
      </div>
      <div
        class="Box"
      >
        <div
          class="Box-box"
          style="background-color: red; width: 200px; height: 120px;"
        />
        <button
          class="Box-delete"
          id="2"
        >
          X
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
